### PR TITLE
ddl, meta: check flashback history time range

### LINKED ISSUE #37585: finish flashback to 20:00, should not allow to flashback 20:10 again
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
mysql> flashback cluster as of timestamp '2022-09-03 20:00';
Query OK, 0 rows affected (12.85 sec)

mysql> flashback cluster as of timestamp '2022-09-03 20:10';
Query OK, 0 rows affected (0.59 sec)
```

### 2. What did you expect to see? (Required)
flashback cluster as of timestamp '2022-09-03 20:10'; should return error
### 3. What did you see instead (Required)
return success
### 4. What is your TiDB version? (Required)
tidb patch

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #37585 

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
store flashback history in TiKV, avoid overlapped flashback TS range
```

### CODE REVIEW

File: ddl/cluster.go [hunk 1]
```diff
@@ -491,15 +494,73 @@ func finishFlashbackCluster(w *worker, job *model.Job) error {
...
+			newTsRange = append(newTsRange, tsRange)
+		} else if startTS < tsRange.StartTS {
+			newTsRange = append(newTsRange, meta.TSRange{StartTS: startTS, EndTS: endTS})
+			break
+		} else {
+			// If startTS in range [tsRange.StartTs, tsRange.EndTs], it's an impossible startTS.
+			return errors.Errorf("Invalid flashback ts range, startTS in old time range")
+		}
+	}
+	if len(newTsRange) == 0 || newTsRange[len(newTsRange)-1].EndTS != endTS {
```
Comment (xhebox) [inline ddl/cluster.go:562]: I think there is a simpler method, filter gc unsafe ranges, insert ranges `startTS > tsRange.EndTS`.

And whatever the cases is, append `[startTs, endTs]`.

File: ddl/cluster.go [hunk 2]
```diff
@@ -491,15 +494,73 @@ func finishFlashbackCluster(w *worker, job *model.Job) error {
...
+			continue
+		}
+		if startTS > tsRange.EndTS {
+			newTsRange = append(newTsRange, tsRange)
+		} else if startTS < tsRange.StartTS {
+			newTsRange = append(newTsRange, meta.TSRange{StartTS: startTS, EndTS: endTS})
+			break
+		} else {
+			// If startTS in range [tsRange.StartTs, tsRange.EndTs], it's an impossible startTS.
+			return errors.Errorf("Invalid flashback ts range, startTS in old time range")
```
Comment (xhebox) [inline ddl/cluster.go:559]: Should not happend since `flashback` is cluster-wide locking command, this comment should be more friendly.

Comment (xhebox) [reply ddl/cluster.go:564]: I mean the error, actually, comment will not be logged, but the error will.

File: ddl/cluster.go [hunk 3]
```diff
@@ -491,15 +494,78 @@ func finishFlashbackCluster(w *worker, job *model.Job) error {
...
+	for _, tsRange := range tsRanges {
+		if tsRange.EndTS < gcSafePoint {
+			continue
+		}
+		if startTS > tsRange.EndTS {
+			// tsRange.StartTS < tsRange.EndTS < startTS.
+			// We should keep tsRange in slices.
+			newTsRange = append(newTsRange, tsRange)
+		} else if startTS < tsRange.StartTS {
+			// startTS < tsRange.StartTS < tsRange.EndTS.
```
Comment (xhebox) [inline ddl/cluster.go:558]:
```
// startTS < tsRange.StartTS < tsRange.EndTS < endTS(= new startTS, thus the max)
```

Want something for `endTS`, too.

File: ddl/cluster.go (hunk 2 above)
Comment (xhebox) [reply ddl/cluster.go:564]: In fact, I mean the error, which may be logged or returned.

**Review by xhebox — APPROVED** [review]

File: ddl/cluster.go (hunk 3 above)
Comment (Defined2014) [reply ddl/cluster.go:557]: checked in https://github.com/pingcap/tidb/pull/37737/files#diff-ce4a909a18db5fdccf707860be14b509161216c5041aa76fff6110dfefe9bf86R542

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**